File: v1_17_R1/src/main/java/org/elliotnash/coordsoffline/v1_17_r1/v1_17_R1.java
package org.elliotnash.coordsoffline.v1_17_r1;

import com.mojang.authlib.GameProfile;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.Level;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.elliotnash.coordsoffline.nmsinterface.NmsManager;


public class v1_17_R1 implements NmsManager {
    public Player loadOfflinePlayer(OfflinePlayer player) {
        MinecraftServer minecraftserver = MinecraftServer.getServer();
        GameProfile gameprofile = new GameProfile(player.getUniqueId(), player.getName());
        ServerPlayer entity = new ServerPlayer(
                minecraftserver,
                minecraftserver.getLevel(Level.OVERWORLD),
                gameprofile
        );

        final Player target = entity.getBukkitEntity();
        if (target != null)
            target.loadData();
        return target;
    }
}

File: src/main/kotlin/org/elliotnash/coordsoffline/nms/v1_17_R1.kt
package org.elliotnash.coordsoffline.nms

import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import org.elliotnash.coordsoffline.OfflinePlayerLoader

class v1_17_R1 : OfflinePlayerLoader, v1_17_R1_v1_21_R1_Base() {
    // ServerPlayer(minecraftServer: MinecraftServer, worldServer: ServerLevel, gameProfile: GameProfile)
    private val serverPlayerConstructor = serverPlayerClass.getConstructor(minecraftServerClass, serverLevelClass, gameProfileClass)

    override fun loadOfflinePlayer(offlinePlayer: OfflinePlayer): Player? {
        // new GameProfile(offlinePlayer.uniqueId, offlinePlayer.name)
        val gameProfile = gameProfileConstructor.newInstance(offlinePlayer.uniqueId, offlinePlayer.name)
        // new ServerPlayer(server, level, gameProfile)
        val serverPlayer = serverPlayerConstructor.newInstance(
            minecraftServerInstance,
            // minecraftServerInstance.getLevel(Level.OVERWORLD)
            minecraftServerGetLevelMethod.invoke(minecraftServerInstance, levelOVERWORLDField.get(null)),
            gameProfile
        )

        // serverPlayer.getBukkitEntity()
        val player = serverPlayerGetBukkitEntity.invoke(serverPlayer) as Player?
        player?.loadData()
        return player
    }
}
